test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,501 @@
+    luatest.assert(ok, string.format(
+        'Promoting unconfigured server is always successful: %s', err))
+
+    local ok, err = pcall(box.ctl.demote)
+    luatest.assert(ok, string.format(
+        'Demoting unconfigured server is always successful: %s', err))
+end
+
+-- Promoting current RAFT leader and synchro queue owner should succeed
+-- with elections enabled.
+g_common.test_leader_promote = function(g)
+    box_cfg_update({g.server_1}, {election_mode = 'manual'})
+    promote(g.server_1)
+    wait_sync(g.cluster.servers)
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'Promoting current leader with elections on succeeds: %s', err))
+    wait_sync(g.cluster.servers)
+
+    box_cfg_update({g.server_1}, {election_mode = 'off'})
+    promote(g.server_1)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'Promoting current leader with elections off succeeds: %s', err))
+    wait_sync(g.cluster.servers)
+
+    demote(g.server_1)
+end
+
+-- Demoting current follower should succeed.
+g_common.test_follower_demote = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'Demoting follower with elections off is always successful: %s', err))
+    wait_sync(g.cluster.servers)
+
+    box_cfg_update({g.server_1}, {election_mode = 'manual'})
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'Demoting follower with elections on is always successful: %s', err))
+end
+
+-- Promoting current RAFT leader should succeed,
+-- even if he doesn't own synchro queue with elections enabled.
+g_common.test_raft_leader_promote = function(g)
+    box_cfg_update({g.server_1}, {election_mode = 'manual'})
+
+    -- Promote server, but get stuck while writing PROMOTE
+    -- (become RAFT leader without obtaining synchro queue)
+    wal_delay_start(g.server_1, 0)
+    local f = promote_start(g.server_1)
+    play_wal_until_synchro_queue_is_busy(g.server_1)
+    g.server_1:wait_election_leader()
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'Promoting current RAFT leader succeeds: %s', err))
+
+    wal_delay_end(g.server_1)
+    fiber_join(g.server_1, f)
+    wait_sync(g.cluster.servers)
+
+    box_cfg_update({g.server_1}, {election_mode = 'off'})
+    demote(g.server_1)
+end
+
+-- Promoting and demoting should work when everything is ok.
+g_common.test_ok = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'Promoting succeeds with elections off: %s', err))
```

Comment:
This place flaky fails when I run the test a lot:
```
[031] Test failed! Output from reject file /tmp/t/rejects/replication-luatest/gh_6033_box_promote_demote.reject:
[031] TAP version 13
[031] 1..12
[031] # Started on Thu Jun 16 21:20:25 2022
[031] # Starting group: gh-6033-box-promote-demote
[031] ok     1	gh-6033-box-promote-demote.test_leader_promote
[031] ok     2	gh-6033-box-promote-demote.test_follower_demote
[031] ok     3	gh-6033-box-promote-demote.test_raft_leader_promote
[031] not ok 4	gh-6033-box-promote-demote.test_ok
[031] #   .../replication-luatest/gh_6033_box_promote_demote_test.lua:224: Promoting succeeds with elections off: Instance with replica id 1 was promoted first
[031] #   expected: a value evaluating to true, actual: false
[031] #   stack traceback:
[031] #   	.../replication-luatest/gh_6033_box_promote_demote_test.lua:224: in function 'gh-6033-box-promote-demote.test_ok'
[031] #   	...
[031] #   	[C]: in function 'xpcall'
[031] ok     5	gh-6033-box-promote-demote.test_simultaneous
[031] ok     6	gh-6033-box-promote-demote.test_voter_promote
[031] ok     7	gh-6033-box-promote-demote.test_wal_interfering_promote
[031] ok     8	gh-6033-box-promote-demote.test_wal_interfering_demote
[031] ok     9	gh-6033-box-promote-demote.test_limbo_full_interfering_promote
[031] ok     10	gh-6033-box-promote-demote.test_limbo_full_interfering_demote
[031] ok     11	gh-6033-box-promote-demote.test_fail_limbo_ack_promote
[031] # Starting group: gh-6033-box-promote-demote-unconfigured
[031] ok     12	gh-6033-box-promote-demote-unconfigured.test_unconfigured
[031] # Ran 12 tests in 2.097 seconds, 11 succeeded, 1 failed

```

Reply:
I guess a `wait_sync` is missing before the promotion?

Reply:
I've tried my best to reproduce, on my machine i didn't fail once in like 5k runs:(. Adding wait_sync shouldn't help, because there is a wait_sync in the end of before_each. But you can try it. Also you can try the following patch:
<details>
<summary>patch</summary>

```diff
diff --git a/test/replication-luatest/gh_6033_box_promote_demote_test.lua b/test/replication-luatest/gh_6033_box_promote_demote_test.lua
index 81523de47..1a88e7934 100644
--- a/test/replication-luatest/gh_6033_box_promote_demote_test.lua
+++ b/test/replication-luatest/gh_6033_box_promote_demote_test.lua
@@ -22,15 +22,39 @@ local function wait_sync(servers)
     end
 end

+-- Run box.ctl.promote until server is RAFT leader and limbo owner.
+-- Might need to run several promotes because of some error. After this
+-- function returns we can be 100% sure, that server considers itself leader.
+-- After that wait until server is RW.
 local function promote(server)
-    server:exec(function()
-        box.ctl.promote()
-        box.ctl.wait_rw()
+    luatest.helpers.retrying({}, function()
+        return server:exec(function()
+            pcall(box.ctl.promote)
+            if box.info.election.state ~= 'leader' then
+                return false
+            end
+            if box.info.synchro.queue.owner ~= box.info.id then
+                return false
+            end
+            return true
+        end)
     end)
+    server:exec(function() box.ctl.wait_rw() end)
 end

+-- Run box.ctl.demote until server is not RAFT leader.
+-- Server might still be RW and/or limbo owner, if elections are turned off.
+-- See https://github.com/tarantool/tarantool/issues/6860
 local function demote(server)
-    server:exec(function() box.ctl.demote() end)
+    luatest.helpers.retrying({}, function()
+        return server:exec(function()
+            pcall(box.ctl.demote)
+            if box.info.election.state == 'leader' then
+                return false
+            end
+            return true
+        end)
+    end)
 end

 -- Allow WAL writes one by one until the synchro queue becomes busy. The
```
</details>

Sometime ago promoting and demoting like that solved similar problems.

Reply:
Thanks @sergepetrenko for investigating this problem. Found out that previous promote from same instance from previous test was interfering with promote in test_ok. Fixed. Also added waiting for new synchro queue term in similar place.

<details>
<summary>fix</summary>

```diff
diff --git a/test/replication-luatest/gh_6033_box_promote_demote_test.lua b/test/replication-luatest/gh_6033_box_promote_demote_test.lua
index 81523de47..d3a48a046 100644
--- a/test/replication-luatest/gh_6033_box_promote_demote_test.lua
+++ b/test/replication-luatest/gh_6033_box_promote_demote_test.lua
@@ -198,6 +198,7 @@ g_common.test_raft_leader_promote = function(g)
     -- Promote server, but get stuck while writing PROMOTE
     -- (become RAFT leader without obtaining synchro queue)
     wal_delay_start(g.server_1, 0)
+    local term = g.server_1:election_term()
     local f = promote_start(g.server_1)
     play_wal_until_synchro_queue_is_busy(g.server_1)
     g.server_1:wait_election_leader()
@@ -210,6 +211,7 @@ g_common.test_raft_leader_promote = function(g)

     wal_delay_end(g.server_1)
     fiber_join(g.server_1, f)
+    g.server_1:wait_synchro_queue_term(term + 1)
     wait_sync(g.cluster.servers)

     box_cfg_update({g.server_1}, {election_mode = 'off'})
```
</details>